File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        def memoize(func):
            cache = {}
            def wrapper(*args):
                if args in cache:
                    return cache[args]
                result = func(*args)
                cache[args] = result
                return result
            return wrapper
        
        
        n = len(s)
        
        @memoize
        def dp(left: int, right: int) -> int:
            if left > right:
                return 0
            if left == right:
                return 1
            
            if s[left] == s[right]:
                return dp(left + 1, right - 1) + 2
            else:
                return max(dp(left, right - 1), dp(left + 1, right))
        
        return dp(0, n - 1)

File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.java
import java.util.*;

class Solution {
    private Map<String, Integer> memo = new HashMap<>();
    
    public int longestPalindromeSubseq(String s) {
        return dp(0, s.length() - 1, s);
    }
    
    private int dp(int left, int right, String s) {
        if (left > right) {
            return 0;
        }
        if (left == right) {
            return 1;
        }
        
        String key = left + "," + right;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        
        int result;
        if (s.charAt(left) == s.charAt(right)) {
            result = dp(left + 1, right - 1, s) + 2;
        } else {
            result = Math.max(dp(left, right - 1, s), dp(left + 1, right, s));
        }
        
        memo.put(key, result);
        return result;
    }
}

